app/id_generation.py

```diff
@@ -2,41 +2,143 @@
 import requests
 from settings import settings
 import random
+from helpers import db_request_token, is_response_valid, is_response_empty
+from request import build_request
+from router import (
+    user as user_router,
+    student as student_router,
+    school,
+    enrollment_record,
+)
 
 school_db_url = settings.db_url + "/school"
 
 
-class JNVIDGeneration:
-    def __init__(self, region, school, grade):
-        self.region = region
-        self.school = school
-        self.grade = grade
-        self.id = (
-            self.get_class_code()
-            + self.get_jnv_code()
-            + self.generate_three_digit_code()
-        )
+async def JNV_ID_generation(parameters):
+    record_already_exist, ID = dedupe_for_users(parameters)
+
+    if record_already_exist:
+        return ID
+    else:
+        counter = 1000
+        while counter > 0:
+            id = (
+                get_class_code(parameters["grade"])
+                + get_jnv_code(parameters["region"], parameters["school_name"])
+                + generate_three_digit_code()
+            )
+
+            duplicate_id = await check_for_duplicate_ID(id)
+            if not duplicate_id:
+                return id
+                break
+            else:
+                counter -= 1
+    raise HTTPException(
+        status_code=500, detail="ID could not be generated, max loops hit!"
+    )
 
-    def get_id(self):
-        return self.id
 
-    def get_class_code(self):
-        class_codes = {"9": "26", "10": "25", "11": "24", "12": "23"}
-        return class_codes[self.grade]
+def dedupe_for_users(parameters):
 
-    def get_jnv_code(self):
-        response = requests.get(
-            school_db_url, params={"region": self.region, "name": self.school_name}
+    # check if the keys used for deduping exist in the request data
+    for key in ["grade", "date_of_birth", "school_name", "gender", "category"]:
+        if key not in parameters or parameters[key] == "" or parameters[key] is None:
+            raise HTTPException(
+                status_code=400,
+                detail="{} is not part of the request data".format(key),
+            )
+
+    does_user_already_exist = user_router.get_users(
```

Comment:
This whole logic exist on the `db-service` end. Will avoid multiple calls to DB service and you can just get the successful ID